Ömer Gözüaçık, 29/05/2020

In [1]:
import pandas as pd
import numpy as np

from sklearn.metrics import mean_squared_error
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_validate
from scipy.optimize import minimize

import warnings 
warnings.filterwarnings('ignore')

## Importing data

In [2]:
pd.set_option('display.max_columns',None)

# training data
train = pd.read_csv('train.csv')

# test data
test = pd.read_csv('test.csv')
df=pd.concat([train,test], sort=False)
train.head()

galactic year                        galaxy  existence expectancy index  \
0         990025  Large Magellanic Cloud (LMC)                    0.628657   
1         990025              Camelopardalis B                    0.818082   
2         990025                       Virgo I                    0.659443   
3         990025            UGC 8651 (DDO 181)                    0.555862   
4         990025                  Tucana Dwarf                    0.991196   

   existence expectancy at birth  Gross income per capita  Income Index  \
0                      63.125200             27109.234310      0.646039   
1                      81.004994             30166.793958      0.852246   
2                      59.570534              8441.707353      0.499762   
3                      52.333293                      NaN           NaN   
4                      81.802464             81033.956906      1.131163   

   Expected years of education (galactic years)  \
0                                      8.240543   
1                                     10.671823   
2                                      8.840316   
3                                           NaN   
4                                     13.800672   

   Mean years of education (galactic years)  \
0                                       NaN   
1                                  4.742470   
2                                  5.583973   
3                                       NaN   
4                                 13.188907   

   Intergalactic Development Index (IDI)  Education Index  \
0                                    NaN              NaN   
1                               0.833624         0.467873   
2                               0.469110         0.363837   
3                                    NaN              NaN   
4                               0.910341         0.918353   

   Intergalactic Development Index (IDI), Rank  \
0                                          NaN   
1                                   152.522198   
2                                   209.813266   
3                                          NaN   
4                                    71.885345   

   Population using at least basic drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Population using at least basic sanitation services (%)  \
0                                                NaN         
1                                                NaN         
2                                                NaN         
3                                                NaN         
4                                                NaN         

   Gross capital formation (% of GGP)  Population, total (millions)  \
0                           38.305483                    681.841086   
1                           36.255559                    423.973089   
2                           18.093991                    285.572207   
3                                 NaN                    327.202247   
4                           34.497468                    633.799718   

   Population, urban (%)  Mortality rate, under-five (per 1,000 live births)  \
0              22.131436                                         150.259290    
1              89.294181                                         117.069332    
2              57.491083                                         226.523760    
3              47.785456                                         279.000175    
4             103.562629                                          28.359587    

   Mortality rate, infant (per 1,000 live births)  \
0                                      129.841789   
1                                       6

### Converting categorical feature to numbers

There is no specific reason. I think it is easier for the eye.

In [3]:
df["galaxy"] = df["galaxy"].astype('category')
df["galaxy"] = df["galaxy"].cat.codes
train = df[:3865]
test = df[3865:]
test=test.drop("y", axis = 1)
test_res= test.copy()

### Checking how many galaxies are there and how many of them are distinct.

- There are **181** distinct galaxies on the training set and **172** on the test set.

- On overall they each galaxy has **20** samples on the training set and **5** on the test set.

- **Some galaxies on the training set does not exist on the test set.**

- **Galaxy 126** has only one sample. I discard it on the training phase

As far as I know, the world bank has **182** members (countries) in 2000s (IBRD). Each distinct galaxy may represent a country in real life. Every sample for a galaxy may represent the properties of the country at a time (galactic year). 

In [4]:
train_gal=set(train["galaxy"])
s=0
for x in train_gal:
    s=s+len(train.loc[train['galaxy'] == x])
print("Total distinct galaxies: {}".format(len(train_gal)))
print("Average samples per galaxy: {}".format(s/len(train_gal)))

Total distinct galaxies: 181
Average samples per galaxy: 21.353591160220994


In [5]:
test_gal=set(test["galaxy"])
s=0
for x in test_gal:
    s=s+len(test.loc[test['galaxy'] == x])
print("Total distinct galaxies: {}".format(len(test_gal)))
print("Average samples per galaxy: {}".format(s/len(test_gal)))

Total distinct galaxies: 172
Average samples per galaxy: 5.174418604651163


#### Number of samples and features
Train set: 3865

Test set: 890

Features: 79

In [6]:
print("Train vector: " + str(train.shape))
print("Test vector: " + str(test.shape))

Train vector: (3865, 80)
Test vector: (890, 79)


## Methods for Cross-validating Training Data

- I trained **a model for exery distinct galaxy** in the training set (180) except the one from 126th galaxy as it has only one sample. 

- I used **features with top x correlation** with respect to y (target variable) galaxy specific. (x is found by trying different values [20,25,30,40,50,60,70])

- Missing values are filled with the galaxy specific 'mean' of the data. (Median can be used alternatively.)

- **Train and test sets are not mixed for both imputation and standardization.**

- Standard Scaler is used to standardize data.

- Gradient Boosted Regression is used as a model.

In [7]:
def cross_validation_loop(data,cor):
    labels= data['y']
    data=data.drop('galaxy', axis=1)    
    data=data.drop('y', axis=1)
    
    correlation=abs(data.corrwith(labels))
    columns=correlation.nlargest(cor).index
    data=data[columns]
    
    imp = SimpleImputer(missing_values=np.nan, strategy='mean').fit(data)
    data=imp.transform(data)

    scaler = StandardScaler().fit(data)
    data = scaler.transform(data)
        
    estimator = GradientBoostingRegressor(n_estimators=300)
    
    cv_results = cross_validate(estimator, data, labels, cv=4, scoring='neg_root_mean_squared_error')

    error=np.mean(cv_results['test_score'])
    
    return error

#### Code for cross-validating a model for every galaxy

I return the mean of the cross-validation scores disregarding the differences of their sample sizes.

In [8]:
train_gal=set(train["galaxy"])
train_gal.remove(126)
def loop_train(cor):
    errors=[]
    for gal in train_gal:
        index = train.index[train['galaxy'] == gal]
        data = train.loc[index]
        errors.append(cross_validation_loop(data,cor))
    return np.mean(errors)

#### Checking which correlation threshold gives better value

The model performs best when the threshold is 20 with RMSE of 0.0063

In [9]:
cor=[20,25,30,40,50,60,70,80]
errors=[]
for x in cor:
    errors.append(loop_train(x))

In [10]:
print(errors)

[-0.006323668899261232, -0.006380895115984182, -0.006426245069277023, -0.006545573360409307, -0.006794901579802094, -0.006924152600585917, -0.007016793213293234, -0.0070351788842627095]


## Making predictions on the test data

- Similar methodology is used to fill the missing value and standardization.
- The best covariance threshold in the cross validation, 20, is used.

In [11]:
def test_loop(data, test_data):
    labels= data['y']
    data=data.drop('galaxy', axis=1)    
    data=data.drop('y', axis=1)
    correlation=abs(data.corrwith(labels))
    columns=correlation.nlargest(20).index
    
    train_labels= labels
    train_data=data[columns]
    test_data= test_data[columns]
    
    imp = SimpleImputer(missing_values=np.nan, strategy='mean').fit(train_data)
    train_data=imp.transform(train_data)
    test_data=imp.transform(test_data)

    scaler = StandardScaler().fit(train_data)
    train_data = scaler.transform(train_data)
    test_data = scaler.transform(test_data)

    model = GradientBoostingRegressor(n_estimators=300)
    model.fit(train_data, train_labels)

    predictions = model.predict(test_data)
    return predictions

#### Sorting samples with respect to their unique galaxy type. 

In [12]:
test=test_res
test=test.sort_values(by=['galaxy'])
test_pred = pd.DataFrame(0, index=np.arange(len(test)), columns=["predicted_y"])

#### Looping over all galaxy types in the test set and making predictions.

In [13]:
i=0
for gal in test_gal:
    count=len(test.loc[test['galaxy'] == gal])
    index = train.index[train['galaxy'] == gal]
    data = train.loc[index]
    pred=test_loop(data,test.loc[test['galaxy']==gal])
    test_pred.loc[i:i+count-1,'predicted_y'] = pred
    i=i+count 

#### Sorting samples with respect to the index.

In [14]:
test["predicted_y"]=test_pred.to_numpy()
test.sort_index(inplace=True)
predictions = test["predicted_y"]

## Discussion 1

- With this approach, we are **not using 8 galaxies in the training set as they are not in the test set.** (Almost 160 samples)

- A better approach should use them as well.

- According to our theory, every galaxy represent a country and samples are its properties at a time (maybe galactic year represents time).

- Some countries may have missing values as they may have joined IBRD late. This may be organizers decision as well. Filling missing values with regression can improve performance.

- World Bank categorizes countries by both region and income: https://datahelpdesk.worldbank.org/knowledgebase/articles/906519-world-bank-country-and-lending-groups

7 regions: East Asia and Pacific, Europe and Central Asia, Latin America & the Caribbean, Middle East and North Africa, North America, South Asia, Sub-Saharan Africa

4 income groups: Low-income economies, Lower-middle-income economies, Upper-middle-income economies, High-income economies 

- Clustering galaxies may excel the performance of the model. I would try both clustering galaxies to either 4 or 7 clusters. Then try making imputation/training with respect to every cluster.

This code is a summary of what we have done. We also analyzed RMSE for cross-validation for per galaxy. 

Galaxies: {128, 2, 4, 5, 133, 11, 140, 147, 153, 154, 34, 35, 40, 43, 55, 64, 76, 78, 83, 100, 101, 102, 107, 108, 119} have RMSE over 0.008. 

The list gives them in order, 128th having 0.008559 and 119th having 0.034926. 

- Fine tuning these problematic galaxies with low cross-validation scores can excel the performance of the model

## Optimization part

- Ideally giving 100 to top 500 samples with highest p^2 values should optimize the likely increase.
- However, as the predictions can be faulty, this approach would result with lower Leaderboard Score.

E.g: If the original p^2 value is higher than the predicted p^2, it will increase the error as we are directly giving it 0.

- That's why, I believe its better to spread the risk for the samples in the bordering regions (400< [rank of p^2] <600).
- I assign 100 energy to top 400 samples and 50 energy to the remaining top 200 samples.

In [27]:
index = predictions
pot_inc = -np.log(index+0.01)+3

In [28]:
p2= pot_inc**2

In [29]:
ss = pd.DataFrame({
    'Index':test.index,
    'pred': predictions,
    'opt_pred':0,
    'eei':test['existence expectancy index'], # So we can split into low and high EEI galaxies
})

In [30]:
ss.loc[p2.nlargest(400).index, 'opt_pred']=100
ss=ss.sort_values('pred')
ss.iloc[400:600].opt_pred = 50
ss=ss.sort_index()

In [31]:
increase = (ss['opt_pred']*p2)/1000

In [32]:
print(sum(increase), ss.loc[ss.eei < 0.7, 'opt_pred'].sum(), ss['opt_pred'].sum())

1788.4031542395505 6500 50000


In [33]:
ss[['Index', 'pred', 'opt_pred']].to_csv('submission.csv', index=False)

## Discussion 2

- Optimization can be done better by changing the spreading the risk part (assigning energy to the 400<p^2<600 region).

- You can give values that are decreasing starting from 400th to 600th (99, 98, 97...). 

It is less likely for the 400th sample to be out of top 500, and similarly it is less likely for the 600th sample to be in the top 500. That's why, you can give more energy to the ones in the 400-500 region and less to the 500-600.

- This approach got me and my friend to the best score of 0.04271993 which is ranked 22nd right now.

- As we are out of top 20 and reached the upload limit, we are sharing our approach publicly to help other teams that have worse results. 
